File: frontend/types/dictionary.ts
/** Pitch accent pattern. */
export interface PitchPattern {
  kanji: string
  pattern: string
}

/** Word sense/meaning from dictionary. */
export interface Sense {
  glosses: string[]
  pos: string[]
  misc: string[]
}

/** Dictionary entry. */
export interface DictionaryEntry {
  id: number
  kanji: string[]
  readings: string[]
  senses: Sense[]
  pitch_accent: PitchPattern[]
}

/** Response from dictionary lookup API. */
export interface LookupResponse {
  query: string
  count: number
  entries: DictionaryEntry[]
}

/** Response from pitch accent lookup API. */
export interface PitchLookupResponse {
  reading: string
  count: number
  patterns: PitchPattern[]
}

File: backend/app/schemas/dictionary.py
"""Schemas for dictionary API responses."""

from pydantic import BaseModel, Field


class Sense(BaseModel):
    """Word sense/meaning."""

    glosses: list[str] = Field(default_factory=list)
    pos: list[str] = Field(default_factory=list)
    misc: list[str] = Field(default_factory=list)


class PitchPattern(BaseModel):
    """Pitch accent pattern."""

    kanji: str = ""
    pattern: str = ""


class DictionaryEntry(BaseModel):
    """Single dictionary entry."""

    id: int
    kanji: list[str] = Field(default_factory=list)
    readings: list[str] = Field(default_factory=list)
    senses: list[Sense] = Field(default_factory=list)
    pitch_accent: list[PitchPattern] = Field(default_factory=list)


class LookupResponse(BaseModel):
    """Response for dictionary lookup."""

    query: str
    count: int
    entries: list[DictionaryEntry]


class PitchLookupResponse(BaseModel):
    """Response for pitch accent lookup."""

    reading: str
    count: int
    patterns: list[PitchPattern]
